### PR TITLE
Feature: Exclude search terms

### PR DESCRIPTION
**Pull request by Jesulayomy**
# Exclude search terms and default configurations

## Exclude search terms

Most of the classes includes this to implement excludes
```python
...
if config and "exclude_search_terms" in config:
    lower_title = job.text.lower()
    if any(
        term in lower_title for term in set([excluded_word.lower() for excluded_word in config["exclude_search_terms"]])
    ):
        continue
...
```

## Default Config
Added a default_config which takes the same parameters as config. At the moment only exclude_search_terms is used.
a company config will override the default config, so you can set a single default and disable / customize it for a company by adding config to that company. 

```python
if "config" not in company:
    company["config"] = default_config
```

```json
{
	"search_terms": ["software", "developer", "engineer"],
	"companies": [
		{
			"name": "Athena Health",
			... // Will use default_config by default
		},
		{
			"name": "Capital One",
			...,
			"config": {} // Disables default config options in Capital one
		},
	],
	"default_config": {
		"exclude_search_terms": [ "intern", "(m/f/d)", "(f/m/d)", "(m/w/d)", "President"],
		...
	}
}
```

### Affected Classes

This PR implements the exclude search terms for various scraper classes
- [x] GreenhousePage
- [x] GreenhouseEmbededStandalonePage
- [x] LeverCoPage
- [x] BambooPage
- [x] WorkablePage
- [x] WorkdayPage
- [x] RipplingPAge
- [x] AshbyPage
- [x] ApplyToJobPage
- [x] SmartRecruitersPage
- [x] BitsInBioPage

### CODE REVIEW

File: example_files/config.json [hunk 1]
```diff
@@ -33,5 +33,8 @@
 			"referral": "",
 			"application_history": ""
 		}
-	]
+	],
+	"default_config": {
+		"exclude_search_terms": [ "intern", "(m/f/d)", "(f/m/d)", "(m/w/d)", "QA", "Senior", "Sr.", "Sr", "Lead", "Deployed", "Principal", "Founding", "Servicenow", "splunk"]
```
Comment (jo-room) [inline example_files/config.json]: nit: Could you modify this to just be the first couple? And not hardcode senior/specific companies into the repo 😛

File: common_scrapers.py [hunk 2]
```diff
@@ -318,6 +379,12 @@ def get_jobs(driver, config=None):
         job_elements = container.find_elements(By.CLASS_NAME, "job-item")
         jobs = []
         for job in job_elements:
+            if config and "exclude_search_terms" in config:
+                lower_title = job.text.lower()
+                if any(
+                    term in lower_title for term in set([excluded_word.lower() for excluded_word in config["exclude_search_terms"]])
```
Comment (jo-room) [inline common_scrapers.py]: This PR implements all of these as case insensitive checks, but your example `exclude_search_terms` includes strings like `"QA"` and `"Sr"`. The case insensitivity would filter out any job title with "qa" or "sr", e.g. maybe if the company/team/location was named Qantas/Srixon/Israel. I think these should be case sensitive, and the configurer can over-specify multiple capitalization combinations if needed.

Side nits, that don't block the PR and I'll follow up on at some point if you don't (no worries about these, try not to procrastinate by coding 😜):
- Instead of specifying `job.text` twice both here when passing into `JobPosting`, and therefore causing the risk of them diverging, I think `job.text` for these should be refactored to a `title = job.text` that is reused in both places
- This carries over from the first ClimateTechList scraper that implemented, but I think there doesn't need to be a set in `set([excluded_word.lower() for excluded_word in config["exclude_search_terms"]])`, it can just be a list that isn't coerced into a set.

**Review by jo-room — CHANGES_REQUESTED** [review]
Sweet, thanks! Blocking on the case sensitivity and nit about not including e.g. splunk, but otherwise LGTM :)

File: example_files/config.json (hunk 1 above)
Comment (Jesulayomy) [reply example_files/config.json]: Done ✅

File: common_scrapers.py (hunk 2 above)
Comment (Jesulayomy) [reply common_scrapers.py]: I think a regex could match them better

- [x] I have a function defined at the top to handle this to avoid repeating the code.
```python
def shouldExclude(config, title) -> bool
```

- [x] Also used the `title=job.text` or `job["text"]`

File: common_scrapers.py [hunk 3]
```diff
@@ -1,27 +1,35 @@
...
 from selenium.webdriver.common.action_chains import ActionChains
 
 
+def is_excluded(config, title) -> bool:
+    if config and "exclude_search_terms" in config:
+        lower_title = title.lower()
+        exclude_terms = [excluded_word.lower() for excluded_word in config["exclude_search_terms"]]
+        pattern = r'\b(?:' + '|'.join(re.escape(term) for term in exclude_terms) + r')\b'
+        if re.search(pattern, lower_title):
+            return True
```
Comment (jo-room) [inline common_scrapers.py]: Blocking: So despite the `re.escape()`, this regex does not match things that contain regex symbols. E.g., if you run the following, it won't return a match (at least on python 3.12.3)

```py
pattern = r'\b(?:' + '|'.join(re.escape(term) for term in ["(m/f/d)"]) + r')\b'
re.search(pattern, "test (m/f/d)")
```

If you did want to match on lowercase, how about matching by word boundaries? It would also be more understandable than regex. As in:

```py
title_words_lower = set(title.lower().split())
exclude_terms = {excluded_word.lower() for excluded_word in config["exclude_search_terms"]}
if len(title_words_lower.intersection(exclude_terms)) > 0:
	return True
```

Note that `.split()` with no args splits on arbitrary whitespace: https://stackoverflow.com/questions/8113782/split-string-on-whitespace-in-python#8113787

**Review by jo-room — CHANGES_REQUESTED** [review]

File: common_scrapers.py (hunk 3 above)
Comment (Jesulayomy) [reply common_scrapers.py]: Thanks! I didn't notice that (Should have tried more test cases) Intersection works okay and yes I will keep the lowercase since the words can appear differently depending on the sites. I guess some edge cases would still break (i.e having `"m / f / d"` would not match since we are splitting on the spaces)

**Review by jo-room — APPROVED** [review]
nit: `import re` is no longer needed, but I'll clean it up and update the Avature scrape on main later

I'm okay with not being able to exclude multi-space words
